### PR TITLE
Use deterministic Process.StartTime for OS X #39434

### LINKED ISSUE #39434: Process.StartTime on macOS returns different values
When getting the StartTime for a process, each time it's called for a new instance of the same process, it has a different value. I believe that the issue is in StartTimeCore where the native "now" time is retrieved, a bunch of arithmetic is performed, then the C# "now" time is retrieved. Since operations were between the two "now" calls, the "now" values will obviously be different, but they're used as if they're the same. I've attached a sample program that will show the deltas between StartTime accesses.

System: macOS 10.14.5 Mojave

[ProcessStartTime.zip](https://github.com/dotnet/corefx/files/3387833/ProcessStartTime.zip)

### PR DESCRIPTION
**Pull request by watfordgnf**
This fixes #39434 by using a deterministic value for `Process.StartTime` from the `pbi_start_*` values returned by `GetProcessInfoById` on Mac OS X.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Thanks. Whatever the right fix is, we should add a test that simply gets the start time several times for a process (eg., the tests process) to check it doesn't change. I assume that test would periodically fail with the current implementation.

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 1]
```diff
@@ -359,6 +359,17 @@ public void StartTime_GetNotStarted_ThrowsInvalidOperationException()
             Assert.Throws<InvalidOperationException>(() => process.StartTime);
         }
 
+        [Fact]
+        public void StartTimeIsStable()
+        {
+            DateTime startTime = Process.GetCurrentProcess().StartTime;
+
+            Thread.Sleep(TimeSpan.FromSeconds(5));
```
Comment (krwq) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: this is a bit long for regular test. 0.5s at most or move it to Outerloop

Comment (watfordgnf) [reply src/System.Diagnostics.Process/tests/ProcessTests.cs]: Can do, I'd picked an arbitrary value less than `WaitInMS` used throughout the tests (30 seconds).

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs [hunk 2]
```diff
@@ -27,37 +29,16 @@ internal DateTime StartTimeCore
...
-                // by some factor “numer / denom”. In most cases the factor is (1 / 1) but we have to get the factor and use it just 
-                // in case we run on a machine has different factor. To get the factor we call GetTimebaseInfo the wrapper for the 
-                // mach_timebase_info() which give us the factor. Then we multiply the factor by the absolute time and the divide
-                // the result by 10^9 to convert it from nanoseconds to seconds.
-
                 EnsureState(State.HaveNonExitedId);
+                Interop.libproc.proc_taskallinfo? info = Interop.libproc.GetProcessInfoById(Id);
 
-                uint numer, denom;
-                Interop.Sys.GetTimebaseInfo(out numer, out denom);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Can any of this interop support be removed now, e.g. can the interop for GetTimebaseInfo be removed?

Comment (watfordgnf) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: A quick GH search says yes. It looks like both GetAbsoluteTime and GetTimebaseInfo are not referenced anywhere other than Process.OSX.

Comment (watfordgnf) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: It goes all the way to System.Native's pal_time.c/pal_time.h. Are these part of corefx, or will this need a coreclr PR as well to remove?

Comment (stephentoub) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: That's all in corefx, e.g. https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Native/pal_time.c

Comment (watfordgnf) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Ok, resected. It appears with that removed `HAVE_MACH_TIMEBASE_INFO` is no longer used as well, so I'll pull that from the cmake config as well.

**Comment by danmosemsft** [comment]
Any more feedback @stephentoub

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs [hunk 3]
```diff
@@ -27,37 +29,16 @@ internal DateTime StartTimeCore
...
-                Interop.libproc.rusage_info_v3 info = Interop.libproc.proc_pid_rusage(_processId);
-                ulong absoluteTime;
+                if (info == null)
+                    throw new Win32Exception(SR.ProcessInformationUnavailable);
 
-                if (!Interop.Sys.GetAbsoluteTime(out absoluteTime))
-                {
-                    throw new Win32Exception(SR.RUsageFailure);
-                }
+                DateTime startTime = DateTime.UnixEpoch + TimeSpan.FromSeconds(info.Value.pbsd.pbi_start_tvsec + info.Value.pbsd.pbi_start_tvusec / MicrosecondsToSecondsFactor);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Should the division include a cast to double, e.g. `pbsd.pbi_start_tvusec / (double)MicrosecondsToSecondsFactor`?

Comment (stephentoub) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Related, have you done any manual testing of this to validate the results?  It looks like our automated testing allows for a lot of wiggleroom on accuracy, around 6 seconds worth.

Comment (watfordgnf) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: For point (1) the factor is already a double. For point (2), I've reached out in the original issue for manual testing beyond CI. 

I'll remove the wiggle room and see what shakes out.

Comment (stephentoub) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: > the factor is already a double

So it is. It's a bit strange that the two factors next to each other have different types.

Comment (watfordgnf) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Disabling the wiggle room apparently fails on multiple platforms (other than OS X).

**Comment by danmosemsft** [comment]
@stephentoub does this address your feedback? If so this can merge.

**Comment by krwq** [comment]
@stephentoub I think this has broken the build. We're getting:

```
##[error]System/Diagnostics/Process.OSX.cs(111,27): error CA1823: Unused field 'NanoSecondToSecondFactor'.
```

in https://github.com/dotnet/corefx/pull/40063

**Comment by krwq** [comment]
I've created https://github.com/dotnet/corefx/pull/40080

**Pull request merged**